### DISCUSSION TITLE
Builtin type incompatibility

### DISCUSSION THREAD

**Discussion by Koswu** [microsoft/pyright#8758] [Q&A]
I have the following code:

```python
from typing import Protocol, Type

class FromStr(Protocol):
    def __init__(self, /, s: str) -> None: ...
    
T: Type[FromStr] = str
```
I expected `str` to match the `FromStr` protocol, but I encountered the following type-checking error:
```
demo.py:8:20 - error: Expression of type "type[str]" cannot be assigned to declared type "type[FromStr]"
    "str" is incompatible with protocol "FromStr"
    Type "type[str]" cannot be assigned to type "type[FromStr]"
      "__init__" is an incompatible type
        Type "() -> None" cannot be assigned to type "(s: str) -> None"
          Function accepts too many positional parameters; expected 0 but received 1 (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Other built-in type, like `int` and `float` have the same problem too.

## Question

Is this behavior expected? Shouldn't `str` be compatible with the `FromStr` protocol since it can be constructed from a `str`?

**Comment by erictraut** [comment]
Yes, this is expected behavior.

According to the [typeshed definition of `str`](https://github.com/python/typeshed/blob/1ace5718deaf3041f8e3d1dc9c9e8a8e830e517f/stdlib/builtins.pyi#L434) (which is the type information pyright uses for builtin types), the class has no `__init__` method but does have a `__new__` method. 

If you want the `str` to be compatible with the `FromStr` protocol, you could change your protocol definition to something like:

```python
class FromStr(Protocol):
    def __new__(cls, object: object = ...) -> Self:
        ...
```

**Reply by Koswu** [reply]
Thanks for reply!

But actually, `str` have an `__init__` method, which can receive any number of arguments 🤔 

```
In [1]: s = "foo"

In [2]: s.__init__('foo', 'bar') # no error

```

**Reply by erictraut** [reply]
If you think the typeshed stubs are incorrect, please file a bug report or PR in the typeshed project.
